src/Features/Core/Portable/PullMemberUp/CodeActionAndSolutionGenerator.cs

```diff
@@ -0,0 +1,197 @@
+            return PullMembersIntoClass(
+                result, codeGenerationService, targetNodeSyntax,
+                solutionEditor, solution, editorMap, syntaxMap);
+        }
+
+        internal async Task<CodeAction> TryGetCodeActionAsync(
+            AnalysisResult result,
+            Document contextDocument,
+            string title)
+        { 
+            var codeGenerationService = contextDocument.Project.LanguageServices.GetRequiredService<ICodeGenerationService>();
+            var targetNodeSyntax = await codeGenerationService.FindMostRelevantNameSpaceOrTypeDeclarationAsync(
+                contextDocument.Project.Solution, result._target).ConfigureAwait(false);
+
+            if (targetNodeSyntax != null)
+            {
+                if (result._target.TypeKind == TypeKind.Interface)
+                {
+                    return PullMembersIntoInterface(result, contextDocument, codeGenerationService, title);
+                }
+                else if (result._target.TypeKind == TypeKind.Class)
+                {
+                    return new SolutionChangeAction(
+                        title,
+                        cancellationToken => GetUpdateSolutionAsync(result, contextDocument, targetNodeSyntax, cancellationToken));
+                }
+                else
+                {
+                    throw ExceptionUtilities.UnexpectedValue(result._target);
+                }
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        private CodeAction PullMembersIntoInterface(
+            AnalysisResult result,
+            Document contextDocument,
+            ICodeGenerationService codeGenerationService,
+            string title)
+        {
+            var options = new CodeGenerationOptions(generateMethodBodies: false, generateMembers: false);
+            var symbolsToPullUp = result._membersAnalysisResults.
+                Select(analysisResult =>
+                {
+                    if (analysisResult._member is IPropertySymbol propertySymbol)
+                    {
+                        // When it is a property, it could have a public getter/setter
+                        // but other one is not public. In this scenario, only the public getter/setter
+                        // will be add the target interface
+                        return CodeGenerationSymbolFactory.CreatePropertySymbol(
+                                propertySymbol,
+                                getMethod: FilterGetterOrSetter(propertySymbol.GetMethod),
+                                setMethod: FilterGetterOrSetter(propertySymbol.SetMethod));
+                    }
+                    else
+                    {
+                        return analysisResult._member;
+                    }
+                });
```

Comment:
why is this work being done out of the lambda?  this just means we do more up front work instead of delaying until when the user invokes the code action.

Reply:
comment still applies.